### DISCUSSION TITLE
Issues - Opening GeoPackage file

### DISCUSSION THREAD

**Discussion by pab2d** [ngageoint/geopackage-android#74] [General]
Hi,

i have a couple of issues opening the created geopackage file, depending on implementation.

Case 1: 

```
final SQLiteDatabase db = featureDao.getDatabaseConnection().getDb();
db.beginTransaction();
try {
	for (ContentValues cv : values) {
		long _id = db.insert(tableName, null, cv);
		if (_id != -1) {
			rowsInserted++;
		}
		db.setTransactionSuccessful();
	}
} finally {
	db.endTransaction();
}
```
Works fast and smooth. The opening inside a client displays an exception:

`Caused by: org.locationtech.jts.io.ParseException: Unknown WKB type 325`

Case 2 - using FeatureRow:

```
for (ContentValues cv : values) {
      FeatureRow featureRow = featureDao.newRow();

	for (Map.Entry<String, Object> entry : cv.valueSet()) {
	    if (geomColumnName.equals(entry.getKey())) {
		    byte[] geometryDataBytes = cv.getAsByteArray(geomColumnName);
		    if (geometryDataBytes != null && geometryDataBytes.length > 0) {
			    GeometryReader gr = new GeometryReader(geometryDataBytes);
			    mil.nga.sf.Geometry geom = null;
			    try {
				    geom = gr.read();
    
				    GeoPackageGeometryData geometryData = new GeoPackageGeometryData(25832);
				    geometryData.setGeometry(geom);
				    featureRow.setGeometry(geometryData);
    
			    } catch (IOException e) {
				    e.printStackTrace();
			    } finally {
				    gr.close();
			    }
		    } else {
			    featureRow.setValue(entry.getKey(), entry.getValue());
		    }
	    }
    }
    
    long _id = featureDao.insert(featureRow);
    if (_id != -1) {
	    rowsInserted++;
    }
    }
}
```
Little slower. Client exception:
```
Caused by: java.sql.SQLException: org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (java.lang.NullPointerException)
	at org.geotools.jdbc.JDBCFeatureReader.runQuery(JDBCFeatureReader.java:282) ~[gt-jdbc-25.2.jar:?]
```

Any ideas how to solve that problem?

**Comment by pab2d** [comment]
Could fix my problem, at least with the case 2 implemenation - had a small bug with wrong brace position. 

But is there a more suitable way?

**Comment by bosborn** [comment]
For case 1 it is hard to tell.  Does the geometry in the ContentValues not have [the required GeoPackage header](https://www.geopackage.org/spec130/#gpb_format)?  This library does not use JTS so I'm not sure the context of your error.  325 is [not a valid geometry type](https://www.geopackage.org/spec130/#geometry_types), so I would guess it is probably corrupt geometry bytes.  Should be an unsigned integer (4 byte) number like [0-3]0[00-17], where the digit in the thousands indicates if z and/or m values are included.

For case 2, you can still use transactions for performance if you wish. See [this test](https://github.com/ngageoint/geopackage-android/blob/3679d011ab1e8f05810f209c97d61cf757945861/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/GeoPackagePerformance.java#L128).
```
FeatureDao dao = ...
dao.beginTransaction();
dao.commit();
dao.endTransaction();
dao.failTransaction();
```

Again I'm not sure the context of your error as this library does not use GeoTools either.  Your code seems to only set geometry column values, not sure if that is intentional or a bug.  Is 25832 a hard coded srs id?  On the geometry data, you could call [setGeometryFromWkb](https://github.com/ngageoint/geopackage-core-java/blob/7456f1df712c05d18436648bccce9690597827d9/src/main/java/mil/nga/geopackage/geom/GeoPackageGeometryData.java#L1606) instead of reading the geometry yourself.  If you are using the GeoPackage through an external connection or library, make sure you close it first to write the changes.

**Reply by pab2d** [reply]
Hi bosborn,
For case 1 your are right, the GP values is missing. 
For case 2, i am already working with transactions (not part of my example) but the Wkb Reader to create the correct header makes it slightly slower (depends on the amount of data). 
The 25832 (UTM 32) is an example SRID and in my case and is just hardcoded.
Thank you

**Reply by bosborn** [reply]
I made [some changes](https://github.com/ngageoint/geopackage-core-java/commit/e36b428974286e9fb5f95553d8d42de7f40a2801) for the next release that should help your situation.

You'll be able to [insert the geometry bytes](https://github.com/ngageoint/geopackage-android/blob/c01553685bb5bc7136d6244c0a62cc6bed7684bf/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/geom/GeoPackageGeometryDataUtils.java#L713) without having to read / parse it into a geometry.  If all your geometries have the same header, you can [determine the header once](https://github.com/ngageoint/geopackage-android/blob/c01553685bb5bc7136d6244c0a62cc6bed7684bf/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/geom/GeoPackageGeometryDataUtils.java#L855) and  [set both the header bytes and geometry bytes ](https://github.com/ngageoint/geopackage-android/blob/c01553685bb5bc7136d6244c0a62cc6bed7684bf/geopackage-sdk/src/androidTest/java/mil/nga/geopackage/geom/GeoPackageGeometryDataUtils.java#L870).

**Reply by bosborn** [reply]
Geometry byte updates released in [version 6.3.0](https://github.com/ngageoint/geopackage-android/releases/tag/6.3.0) and [core version 6.3.0](https://github.com/ngageoint/geopackage-core-java/releases/tag/6.3.0).
